### PR TITLE
Feature determine language

### PR DESCRIPTION
**Pull request by Fasse**
#47

### CODE REVIEW

File: adm_program/system/classes/languagedata.php [hunk 1]
```diff
@@ -73,6 +82,59 @@ public function addLanguagePath($path)
...
+        if(!isset($_SERVER["HTTP_ACCEPT_LANGUAGE"]) || empty($_SERVER["HTTP_ACCEPT_LANGUAGE"]))
+        {
+            return $defaultLanguage;
+        }
+
+        $accepted = preg_split("{,\s*}", $_SERVER["HTTP_ACCEPT_LANGUAGE"]);
+        $language = $defaultLanguage;
+        $quality = 0;
+
+        if(is_array($accepted) && (count($accepted) > 0))
```
Comment (ximex) [inline adm_program/system/classes/languagedata.php]: unnecessary `()`
this is enough: `if(is_array($accepted) && count($accepted) > 0)`

File: adm_program/system/classes/languagedata.php [hunk 2]
```diff
@@ -73,6 +82,59 @@ public function addLanguagePath($path)
     }
 
     /**
+     * Determine the language from the browser preferences of the user.
+     * @return string Return the preferred language code of the client browser
+     */
+    private function determineBrowserLanguage($defaultLanguage)
+    {
+        if(!isset($_SERVER["HTTP_ACCEPT_LANGUAGE"]) || empty($_SERVER["HTTP_ACCEPT_LANGUAGE"]))
```
Comment (ximex) [inline adm_program/system/classes/languagedata.php]: change `"` to `'`

File: adm_program/system/classes/languagedata.php [hunk 3]
```diff
@@ -73,6 +82,59 @@ public function addLanguagePath($path)
...
+        {
+            return $defaultLanguage;
+        }
+
+        $accepted = preg_split("{,\s*}", $_SERVER["HTTP_ACCEPT_LANGUAGE"]);
```
Comment (ximex) [inline adm_program/system/classes/languagedata.php]: change `"` to `'`

File: adm_program/system/classes/languagedata.php [hunk 4]
```diff
@@ -73,6 +82,59 @@ public function addLanguagePath($path)
...
+        if(is_array($accepted) && (count($accepted) > 0))
+        {
+            foreach($accepted as $key => $value)
+            {
+                if(!preg_match("{^([a-z]{1,8}(?:-[a-z]{1,8})*)(?:;\s*q=(0(?:\.[0-9]{1,3})?|1(?:\.0{1,3})?))?$}i", $value, $matches))
+                {
+                    continue;
+                }
+
+                $code = explode("-", $matches[1]);
```
Comment (ximex) [inline adm_program/system/classes/languagedata.php]: change `"` to `'`

File: adm_program/system/classes/languagedata.php [hunk 5]
```diff
@@ -73,6 +82,59 @@ public function addLanguagePath($path)
...
+                else
+                {
+                    $priority = 1.0;
+                }
+
+                while(count($code) > 0)
+                {
+                    if($priority > $quality)
+                    {
+                        $language = strtolower(implode("-", $code));
```
Comment (ximex) [inline adm_program/system/classes/languagedata.php]: change `"` to `'`

File: adm_program/system/classes/languagedata.php [hunk 6]
```diff
@@ -73,6 +82,59 @@ public function addLanguagePath($path)
...
+
+        $accepted = preg_split("{,\s*}", $_SERVER["HTTP_ACCEPT_LANGUAGE"]);
+        $language = $defaultLanguage;
+        $quality = 0;
+
+        if(is_array($accepted) && (count($accepted) > 0))
+        {
+            foreach($accepted as $key => $value)
+            {
+                if(!preg_match("{^([a-z]{1,8}(?:-[a-z]{1,8})*)(?:;\s*q=(0(?:\.[0-9]{1,3})?|1(?:\.0{1,3})?))?$}i", $value, $matches))
```
Comment (ximex) [inline adm_program/system/classes/languagedata.php]: change `"` to `'`

File: adm_program/installation/installation.php [hunk 7]
```diff
@@ -94,10 +94,12 @@
 }
 else
 {
-    $language = 'en';
+    $language = '';
```
Comment (ximex) [inline adm_program/installation/installation.php]: Put this `$language = '';` definition before the if

File: adm_program/system/classes/languagedata.php [hunk 8]
```diff
@@ -43,12 +43,13 @@ class LanguageData
...
      * Therefore the language must be set and optional a path where the language files are stored.
      * @param string $language     The ISO code of the language for which the texts should be read e.g. @b 'de'
+     *                             If no language is set than the browser language will be determined.
      * @param string $languagePath Optional a server path to the language files. If no path is set
      *                             than the default Admidio language path @b adm_program/languages will be set.
      * @param bool   $determineBrowserLanguage If set to true then set browser language of user to the current
      *                             language if possible
      */
-    public function __construct($language, $languagePath = '', $determineBrowserLanguage = true)
+    public function __construct($language = '', $languagePath = '', $determineBrowserLanguage = true)
```
Comment (ximex) [inline adm_program/system/classes/languagedata.php]: param `determineBrowserLanguage` could be removed

**Comment by Fasse** [comment]
@ximex fixed

**Pull request merged**